Sources/MongoSwift/MongoConnectionString.swift

```diff
@@ -0,0 +1,362 @@
+                    // TODO: SWIFT-1163: error on unknown options
+                    break
+                }
+            }
+        }
+
+        private func parseAuthMechanismProperties(properties: String) throws -> BSONDocument {
+            var propertiesDoc = BSONDocument()
+            for property in properties.components(separatedBy: ",") {
+                let kv = property.components(separatedBy: ":")
+                guard kv.count == 2 else {
+                    throw MongoError.InvalidArgumentError(
+                        message: "authMechanismProperties must be a comma-separated list of colon-separated"
+                            + " key-value pairs"
+                    )
+                }
+                switch kv[0].lowercased() {
+                case "service_name", "service_realm":
+                    propertiesDoc[kv[0]] = .string(kv[1])
+                case "canonicalize_host_name":
+                    switch kv[1] {
+                    case "true":
+                        propertiesDoc[kv[0]] = .bool(true)
+                    case "false":
+                        propertiesDoc[kv[0]] = .bool(false)
+                    default:
+                        throw MongoError.InvalidArgumentError(
+                            message: "Value for CANONICALIZE_HOST_NAME in authMechanismProperties must be true or false"
+                        )
+                    }
+                case let other:
+                    throw MongoError.InvalidArgumentError(message: "Unknown key for authMechanismProperties: \(other)")
+                }
+            }
+            return propertiesDoc
+        }
+    }
+
+    /// Parses a new `MongoConnectionString` instance from the provided string.
+    /// - Throws:
+    ///   - `MongoError.InvalidArgumentError` if the input is invalid.
+    public init(throwsIfInvalid input: String) throws {
+        self.description = input
+
+        let schemeAndRest = input.components(separatedBy: "://")
+        guard schemeAndRest.count == 2, let scheme = Scheme(schemeAndRest[0]) else {
+            throw MongoError.InvalidArgumentError(
+                message: "Invalid connection string scheme, expecting \'mongodb\' or \'mongodb+srv\'"
+            )
+        }
+        guard !schemeAndRest[1].isEmpty else {
+            throw MongoError.InvalidArgumentError(message: "Invalid connection string")
+        }
+        let identifiersAndOptions = schemeAndRest[1].components(separatedBy: "/")
+        // TODO: SWIFT-1174: handle unescaped slashes in unix domain sockets.
+        guard identifiersAndOptions.count <= 2 else {
+            throw MongoError.InvalidArgumentError(
+                message: "Connection string contains an unescaped slash"
+            )
+        }
+        let userAndHost = identifiersAndOptions[0].components(separatedBy: "@")
+        guard userAndHost.count <= 2 else {
+            throw MongoError.InvalidArgumentError(message: "Invalid user information")
```

Comment:
I think the error message here could be a little more specific like the one on 213, since it seems like the problem here would likely be an unescaped `@` symbol